### PR TITLE
Validation checks for CMD and ENTRYPOINT

### PR DESCRIPTION
**Pull request by cwebster2**
This checks that CMD and ENTRYPOINT parse as JSON and contain parse as an Array.

It will present the following errors (and opts out of some other checks where these provide more appropriate messaging):

Parses as a string:
image

Does not parse as JSON:
image

ENTRYPOINT present without CMD:
image

### CODE REVIEW

File: checks/entrypoint-requires-cmd.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ */
+async function entrypointRequiresCmd(deployment, context, inputs, httpGet) {
+  /**
+   * You should check the existance of any file you're trying to check
+   */
+  if (!deployment.ordersContents) {
+    log.info(`No Orders Present - Skipping ${deployment.serviceName}`);
+    return [];
+  }
+  log.info(`Template Check - ${deployment.ordersPath}`);
```
Comment (shawnrushefsky) [inline checks/entrypoint-requires-cmd.js:21]:
```suggestion
  log.info(`ENTRYPOINT Requires CMD - ${deployment.ordersPath}`);
```

File: checks/valid-json-arrays-in-bash.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ */
+async function validJsonArrayInBash(deployment, context, inputs, httpGet) {
+  /**
+   * You should check the existance of any file you're trying to check
+   */
+  if (!deployment.ordersContents) {
+    log.info(`No Orders Present - Skipping ${deployment.serviceName}`);
+    return [];
+  }
+  log.info(`Template Check - ${deployment.ordersPath}`);
```
Comment (shawnrushefsky) [inline checks/valid-json-arrays-in-bash.js:22]:
```suggestion
  log.info(`CMD/ENTRYPOINT are valid JSON - ${deployment.ordersPath}`);
```

File: checks/valid-json-arrays-in-bash.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  deployment.ordersContents.forEach((line, i) => {
+    // GitHub lines are 1-indexed
+    const lineNumber = i + 1;
+    // this test only checks CMD and ENTRYPOINT values
+    const r = line.match(/export\s+(CMD|ENTRYPOINT)=(['"])(.*)\2/);
+
+    if (r) {
+      // get value of CMD variable
+      const variableName = r[1];
+      //const quoteStyle = (r[2] == '\'' ? 'single' : 'double');
```
Comment (shawnrushefsky) [inline checks/valid-json-arrays-in-bash.js:36]: remove if unused?

File: checks/valid-json-arrays-in-bash.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+      try {
+        // Is value valid JSON?
+        const obj = JSON.parse(value);
+
+        // it is JSON, but is it an array?
+        if (obj.constructor.name !== "Array") {
+          let suggestion = "";
+
+          if (obj.constructor.name === "String") {
```
Comment (shawnrushefsky) [inline checks/valid-json-arrays-in-bash.js:47]: similar question, why not `typeof obj === "str"`.  Not a request for a change, just curious about why you chose this.

File: checks/valid-json-arrays-in-bash.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            // attempt a solution for them.
```
Comment (shawnrushefsky) [inline checks/valid-json-arrays-in-bash.js:48]:
```suggestion
            // Try to convert string into valid json array
```

File: test/valid-json-arrays-in-bash.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
+const { expect } = require("chai");
+const validJsonArrayInBashCheck = require("../checks/valid-json-arrays-in-bash");
+
+["CMD", "ENTRYPOINT"].map(varName => {
```
Comment (shawnrushefsky) [inline test/valid-json-arrays-in-bash.js:4]: Smart. I've never thought to use `describe` in a `map` like this

**Review by shawnrushefsky — CHANGES_REQUESTED** [review]
Everything looks good except a couple comments and a little cleanup

File: test/valid-json-arrays-in-bash.js (hunk 6 above)
Comment (cwebster2) [reply test/valid-json-arrays-in-bash.js:4]: Thanks!

File: checks/valid-json-arrays-in-bash.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      const variableName = r[1];
+      //const quoteStyle = (r[2] == '\'' ? 'single' : 'double');
+      const value = r[3];
+
+      try {
+        // Is value valid JSON?
+        const obj = JSON.parse(value);
+
+        // it is JSON, but is it an array?
+        if (obj.constructor.name !== "Array") {
```
[the remark this answers is not in the document]
Comment (cwebster2) [reply checks/valid-json-arrays-in-bash.js]: I was curious too, changed to use Array.isArray

**Comment by cwebster2** [comment]
I have addressed all of the comments above, thanks for the input!

**Review by shawnrushefsky — APPROVED** [review]

**Pull request merged**